server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -57,40 +62,110 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
         let instant = Instant::now();
         let ctx = self.handler_ctx();
         let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
-        let handle = self.runtimes.read_runtime.spawn(async move {
-            let read_request = request.into_inner();
-            handle_stream_read(ctx, read_request).await
-        });
+
+        let request = request.into_inner();
+        let table_engine::remote::model::ReadRequest {
+            table,
+            read_request,
+        } = request.clone().try_into().box_err().context(ErrWithCause {
+            code: StatusCode::BadRequest,
+            msg: "fail to convert read request",
+        })?;
+
+        let request_key = RequestKey::new(
+            table.table,
+            read_request.predicate.clone(),
+            read_request.projected_schema.projection(),
+            read_request.order,
+        );
+
+        if let Some(request_notifiers) = self.request_notifiers.clone() {
```

Comment:
I guess we can provide two implementations: `stream_read_internal` & `deduped_stream_read_internal` and  which should be used depends on whether the `request_notifiers` is set.